File: Backend/app/pw_library/python_methods/gauss_seidel.py
import numpy as np


def gauss_seidel(A, b, x0, tol=1e-7, max_iterations=100):
    n = len(b)
    x = np.array(x0, dtype=float)

    for k in range(max_iterations):
        x_old = x.copy()

        for i in range(n):
            s1 = sum(A[i][j] * x[j] for j in range(i))
            s2 = sum(A[i][k] * x_old[k] for k in range(i + 1, n))
            x[i] = (b[i] - s1 - s2) / A[i][i]

        # Check for convergence
        if np.linalg.norm(x - x_old, ord=np.inf) < tol:
            return x, k, abs(np.linalg.norm(x - x_old, ord=np.inf))

    raise ValueError(f'Gauss-Seidel method did not converge after {max_iterations} iterations')

File: Backend/app/pw_library/rust_methods/gauss_seidel.rs
use nalgebra::{DMatrix, DVector};
use std::f64;

// Gauss-Seidel Method
fn gauss_seidel(a: &DMatrix<f64>, b: &DVector<f64>, x0: &DVector<f64>, tol: f64, max_iterations: usize) -> Result<(DVector<f64>, usize, f64), &'static str> {
    let n = b.len();
    let mut x = x0.clone();

    for k in 0..max_iterations {
        let x_old = x.clone();

        for i in 0..n {
            let s1: f64 = (0..i).map(|j| a[(i, j)] * x[j]).sum();
            let s2: f64 = ((i + 1)..n).map(|j| a[(i, j)] * x_old[j]).sum();
            x[i] = (b[i] - s1 - s2) / a[(i, i)];
        }

        // Check for convergence
        let error = (&x - &x_old).norm();
        if error < tol {
            return Ok((x, k, error));
        }
    }

    Err("Gauss-Seidel method did not converge within the maximum number of iterations.")
}
